node/bft/src/sync/mod.rs

```diff
@@ -102,16 +125,67 @@ impl<N: Network> Sync<N> {
         info!("Syncing storage with the ledger...");
 
         // Sync the storage with the ledger.
-        self.sync_storage_with_ledger_at_bootup().await
+        self.sync_storage_with_ledger_at_bootup().await?;
+
+        debug!("Finished initial block synchronization at startup");
+        Ok(())
+    }
+
+    /// Issues new requests for blocks, if needed.
+    ///
+    /// Additionally, this function removes obsolete and timed out block requests,
+    /// and disconnects/bans unresponsive peers.
+    ///
+    /// Responses to block requests will eventually be processed by `Self::try_advancing_block_synchronization`.
+    #[inline]
+    pub async fn issue_block_requests(&self) {
+        // First see if any peers need removal
+        let peers_to_ban = self.block_sync.remove_timed_out_block_requests();
+        for peer_ip in peers_to_ban {
+            trace!("Banning peer {peer_ip} for timing out on block requests");
+
+            let tcp = self.gateway.tcp().clone();
+            tcp.banned_peers().update_ip_ban(peer_ip.ip());
+
+            tokio::spawn(async move {
+                tcp.disconnect(peer_ip).await;
+            });
+        }
+
+        // We might be further ahead than the ledger, if there are queued
+        // responses.
+        let current_height = {
+            let responses = self.latest_block_responses.lock().await;
+            if let Some((height, _)) = responses.last_key_value() { *height } else { self.ledger.latest_block_height() }
+        };
+
+        // Prepare the block requests, if any.
+        // In the process, we update the state of `is_block_synced` for the sync module.
+        let (block_requests, sync_peers) = self.block_sync.prepare_block_requests_at_height(current_height);
+        trace!("Prepared {} block requests", block_requests.len());
+
+        // Sends the block requests to the sync peers.
+        for requests in block_requests.chunks(DataBlocks::<N>::MAXIMUM_NUMBER_OF_BLOCKS as usize) {
+            if !self.block_sync.send_block_requests(&self.gateway, &sync_peers, requests).await {
+                // Stop if we fail to process a batch of requests.
+                break;
+            }
+
+            // Sleep to avoid triggering spam detection.
+            tokio::time::sleep(BLOCK_REQUEST_BATCH_DELAY).await;
+        }
     }
 
     /// Starts the sync module.
+    ///
+    /// When this function returns sucessfully, the sync module will have spawned background tasks
+    /// that fetch blocks from other validators.
     pub async fn run(&self, sync_receiver: SyncReceiver<N>) -> Result<()> {
         info!("Starting the sync module...");
 
         // Start the block sync loop.
         let self_ = self.clone();
-        self.handles.lock().push(tokio::spawn(async move {
```

Comment:
do we not want to keep track of this long-running task? it facilitates a cleaner shutdown

Reply:
ah, it now goes through `self.spawn`, which does the same :+1:

Reply:
Yes, exactly. Not sure why that line did not use self.spawn before, but it was probably just an oversight.